archive/imo/imo1962_q1.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2020 Kevin Lacker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker
+-/
+
+import data.nat.digits
+
+open nat
+
+/-!
+# IMO 1962 Q1
+
+Find the smallest natural number $n$ which has the following properties:
+
+(a) Its decimal representation has 6 as the last digit.
+
+(b) If the last digit 6 is erased and placed in front of the remaining digits,
+the resulting number is four times as large as the original number $n$.
+
+Since Lean does not explicitly express problems of the form "find the smallest number satisfying X",
+we define the problem as a predicate, and then prove a particular number is the smallest member
+of a set satisfying it.
+-/
+
+def problem_predicate (n : ℕ) :=
+(nat.digits 10 n).head = 6 ∧ of_digits 10 ((nat.digits 10 n).tail.concat 6) = 4 * n
+
+/-
+First, it's inconvenient to work with digits, so let's simplify them out of the problem.
+-/
+
+lemma digit_recursion (n : ℕ) (h1 : n > 0) :
+  (nat.digits 10 n) = (n % 10) :: (nat.digits 10 (n / 10)) :=
```

Comment:
This looks like a lemma that should go to `src/`.

Reply:
Certainly it shouldn't be repeated in different IMO files: https://github.com/leanprover-community/mathlib/pull/4366/files#r499114062

There was some discussion about this [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/simplifying.20nat.2Edigits), but I guess there wasn't a clear outcome.

Reply:
My conclusion from the Zulip discussion was that @digama0 figured out a way to make the simplifier work nicely with the digits library, but it was too complicated for me to understand. So I was hoping to punt on it until such time as he submitted that snippet of code he wrote in Zulip.

Reply:
The conversation about this ended up with https://github.com/leanprover-community/mathlib/pull/4452 . Although it's not entirely clear to me whether that would in fact render this lemma unneeded...

Reply:
It looks like that pull request is solving a slightly different problem, so I opened https://github.com/leanprover-community/mathlib/pull/4463 to put this lemma in `data/nat/digits`. Take a look over there, I'll wait on that PR then rewrite this to use it

Reply:
Never mind, this is actually solved by https://github.com/leanprover-community/mathlib/pull/4452 so let's wait on that one.

Reply:
@lacker #4452 is now merged :tada:

Reply:
OK - I wrote this to use the new `digits_def'` and `digit_recursion` is now removed.